src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -1151,131 +1166,132 @@ ObjectDataHandler::read_ret ObjectDataHandler::read(
-        // offset~len falls within reserved region and len > 0
-        ceph_assert(_pins.size() >= 1);
-        ceph_assert(_pins.front().get_key() <= l_start);
-        return seastar::do_with(
-          std::move(_pins),
-          l_start,
-          [FNAME, ctx, l_start, l_end, &ret](auto &pins, auto &l_current) {
-          return trans_intr::do_for_each(
-            pins,
-            [FNAME, ctx, l_start, l_end,
-             &l_current, &ret](auto &pin) -> read_iertr::future<> {
-            auto pin_start = pin.get_key();
-            extent_len_t read_start;
-            extent_len_t read_start_aligned;
-            if (l_current == l_start) { // first pin may skip head
-              ceph_assert(l_current.get_aligned_laddr(
-		ctx.tm.get_block_size()) >= pin_start);
-              read_start = l_current.template
-                get_byte_distance<extent_len_t>(pin_start);
-              read_start_aligned = p2align(read_start, ctx.tm.get_block_size());
-            } else { // non-first pin must match start
-              assert(l_current > l_start);
-              ceph_assert(l_current == pin_start);
-              read_start = 0;
-              read_start_aligned = 0;
-            }
+  LOG_PREFIX(ObjectDataHandler::read);
+  struct read_t : TransactionManager::read_pin_t<ObjectDataBlock> {
+    extent_len_t unaligned_start_offset = 0;
+    extent_len_t unaligned_len = 0;
+    read_t(
+      LBAMapping mapping,
+      extent_len_t aligned_off,
+      extent_len_t aligned_len,
+      extent_len_t unaligned_start_offset,
+      extent_len_t unaligned_len)
+      : read_pin_t(std::move(mapping), aligned_off, aligned_len),
+	unaligned_start_offset(unaligned_start_offset),
+	unaligned_len(unaligned_len) {}
+  };
+  auto ret = bufferlist();
+  auto rpins = std::vector<read_t>();
+  auto guarded_obj_data = guard_object_data(ctx);
+  auto &object_data = guarded_obj_data.object_data;
+  DEBUGT("reading {}~0x{:x}",
+	 ctx.t,
+	 object_data.get_reserved_data_base(),
+	 object_data.get_reserved_data_len());
+  /* Assumption: callers ensure that onode size is <= reserved
+   * size and that len is adjusted here prior to call */
+  ceph_assert(!object_data.is_null());
+  ceph_assert((obj_offset + len) <= object_data.get_reserved_data_len());
+  ceph_assert(len > 0);
+  laddr_offset_t l_start =
+    object_data.get_reserved_data_base() + obj_offset;
+  laddr_offset_t l_end = l_start + len;
+  laddr_t aligned_start = l_start.get_aligned_laddr(
+    ctx.tm.get_block_size());
+  loffset_t aligned_length =
+      l_end.get_roundup_laddr(ctx.tm.get_block_size()).get_byte_distance<
+	loffset_t>(aligned_start);
+  auto _pins = co_await ctx.tm.get_pins(
+    ctx.t,
+    aligned_start,
+    aligned_length);
+  // offset~len falls within reserved region and len > 0
+  ceph_assert(_pins.size() >= 1);
+  ceph_assert(_pins.front().get_key() <= l_start);
+  auto l_current = l_start;
+  for (auto &pin : _pins) {
```

Comment:
using for loop instead of trans_intr::do_for_each may not handle the condition interruption.